File: frontend/src/types/holding.ts
/** Matches backend HoldingResponse (list endpoint, no tags) */
export interface HoldingResponse {
  id: number;
  symbol: string;
  exchange: string;
  instrument_type: string;
  quantity: number;
  average_price: number;
  current_price: number | null;
  current_value: number | null;
  pnl: number | null;
  source: string;
  zerodha_trading_symbol: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

/** Matches backend HoldingWithTags (detail endpoint, includes tags) */
export interface Holding extends HoldingResponse {
  tags: TagBrief[];
}

export interface TagBrief {
  id: number;
  name: string;
  color: string | null;
}

export interface HoldingCreate {
  symbol: string;
  exchange: string;
  instrument_type: string;
  quantity: number;
  average_price: number;
  current_price?: number;
  notes?: string;
}

export interface HoldingUpdate {
  symbol?: string;
  exchange?: string;
  instrument_type?: string;
  quantity?: number;
  average_price?: number;
  current_price?: number;
  notes?: string;
}

File: backend/app/schemas/holding.py
from datetime import datetime

from pydantic import BaseModel


class HoldingBase(BaseModel):
    symbol: str
    exchange: str = "NSE"
    instrument_type: str = "EQ"
    quantity: float
    average_price: float
    current_price: float | None = None
    notes: str | None = None


class HoldingCreate(HoldingBase):
    pass


class HoldingUpdate(BaseModel):
    symbol: str | None = None
    exchange: str | None = None
    instrument_type: str | None = None
    quantity: float | None = None
    average_price: float | None = None
    current_price: float | None = None
    notes: str | None = None


class HoldingResponse(HoldingBase):
    id: int
    current_value: float | None = None
    pnl: float | None = None
    source: str = "manual"
    zerodha_trading_symbol: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class HoldingWithTags(HoldingResponse):
    tags: list["TagBrief"] = []


class TagBrief(BaseModel):
    id: int
    name: str
    color: str | None = None

    model_config = {"from_attributes": True}


class TagAssign(BaseModel):
    tag_ids: list[int]


HoldingWithTags.model_rebuild()
